core/main/src/broker/rules_engine.rs

```diff
@@ -0,0 +1,205 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+use jaq_interpret::{Ctx, FilterT, ParseCtx, RcIter, Val};
+use ripple_sdk::api::{
+    gateway::rpc_gateway_api::RpcRequest, manifest::extn_manifest::ExtnManifest,
+};
+use ripple_sdk::{
+    chrono::Utc,
+    log::{debug, error, info, warn},
+    serde_json::Value,
+    utils::error::RippleError,
+};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::{fs, path::Path};
+
+#[derive(Debug, Deserialize, Default, Clone)]
+pub struct RuleSet {
+    pub endpoints: HashMap<String, RuleEndpoint>,
+    pub rules: HashMap<String, Rule>,
+}
+
+impl RuleSet {
+    pub fn append(&mut self, rule_set: RuleSet) {
+        self.endpoints.extend(rule_set.endpoints);
+        self.rules.extend(rule_set.rules);
+    }
+}
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct RuleEndpoint {
+    pub protocol: RuleEndpointProtocol,
+    pub url: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "lowercase")]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum RuleEndpointProtocol {
+    Websocket,
+    Http,
+    Thunder,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Rule {
+    pub alias: String,
+    pub transform: RuleTransform,
+    pub endpoint: Option<String>,
+}
+
+#[derive(Debug, Clone, Deserialize, Default)]
+pub struct RuleTransform {
+    pub request: Option<String>,
+    pub response: Option<String>,
+    pub event: Option<String>,
+}
+
+impl RuleTransform {
+    pub fn apply_context(&mut self, rpc_request: &RpcRequest) {
+        if let Some(value) = self.request.take() {
+            debug!("Check if value contains {}", value);
+            if value.contains("$context.appId") {
+                debug!("has context");
+                let new_value = value.replace("$context.appId", &rpc_request.ctx.app_id);
+                debug!("changed value {}", new_value);
+                let _ = self.request.insert(new_value);
+            } else {
+                let _ = self.request.insert(value);
+            }
+        }
+    }
+
+    pub fn get_filter(&self, typ: RuleTransformType) -> Option<String> {
+        match typ {
+            RuleTransformType::Request => self.request.clone(),
+            RuleTransformType::Event => self.event.clone(),
+            RuleTransformType::Response => self.response.clone(),
+        }
+    }
+}
+
+pub enum RuleTransformType {
+    Request,
+    Response,
+    Event,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct RuleEngine {
+    pub rules: RuleSet,
+}
+
+impl RuleEngine {
+    fn build_path(path: &str, default_path: &str) -> String {
+        if path.starts_with('/') {
+            path.to_owned()
+        } else {
+            format!("{}{}", default_path, path)
```

Comment:
If `default_path` doesn't end with `/` or path doesn't start with `/` this will not construct the path correctly. I think you can do:

`format!("{}{}", default_path, path)`

...since a path like `/some_directory//some_subdirectory` is still a valid path.

Reply:
Think we avoided that strategy for cases with Windows development environment

Reply:
Regardless of how you want to do it I think you should account for a missing `/`.